pkg/ensign/api/v1beta1/deduplication.go

```diff
@@ -0,0 +1,194 @@
+package api
+
+import (
+	"encoding/binary"
+	"errors"
+	"fmt"
+
+	"github.com/twmb/murmur3"
+	"golang.org/x/exp/slices"
+)
+
+var (
+	ErrNoKeys   = errors.New("no keys specified for key based hashing")
+	ErrNoFields = errors.New("no fields specified for field based hashing")
+)
+
+//===========================================================================
+// Event Hashing Methods
+//===========================================================================
+
+// Hash uses the deduplication policy to determine the hash signature of the event
+// wrapped by the event wrapper and returns the appropriate signature that should be
+// used to detect duplicates in the event stream.
+func (w *EventWrapper) Hash(policy *Deduplication) ([]byte, error) {
+	switch policy.Strategy {
+	case Deduplication_NONE:
+		return nil, nil
+	case Deduplication_STRICT:
+		return w.HashStrict()
+	case Deduplication_DATAGRAM:
+		return w.HashDatagram()
+	case Deduplication_KEY_GROUPED:
+		return w.HashKeyGrouped(policy.Keys)
+	case Deduplication_UNIQUE_KEY:
+		return w.HashUniqueKey(policy.Keys)
+	case Deduplication_UNIQUE_FIELD:
+		return w.HashUniqueField(policy.Keys)
+	default:
+		return nil, fmt.Errorf("unknown deduplication strategy %s", policy.Strategy)
+	}
+}
+
+// Strict hashing is used to detect duplicates where two events have identical metadata,
+// data, mimetype, and type. This method works by setting any non-hash fields to zero
+// values then marshaling the protocol buffers of the event and computing the murmur3
+// hash on the serialized data.
+func (w *EventWrapper) HashStrict() (_ []byte, err error) {
+	var event *Event
+	if event, err = w.Unwrap(); err != nil {
+		return nil, err
+	}
+
+	// Create the hash and write the event data to the hash.
+	hash := murmur3.New128()
+	if _, err = hash.Write(event.Data); err != nil {
+		return nil, fmt.Errorf("could not write event data to hash: %w", err)
+	}
+
+	// Sort the keys in the metadata to write them in lexicographic order
+	keys := make([]string, 0, len(event.Metadata))
+	for key := range event.Metadata {
+		keys = append(keys, key)
+	}
+	slices.Sort(keys)
+
+	// Write the metadata to the to the hash
+	for _, key := range keys {
+		val := event.Metadata[key]
+		if _, err = hash.Write([]byte(key + val)); err != nil {
```

Comment:
Including the key and the value in the meta data hash, concatenated with no separator.